### PR TITLE
With small probability, try an invalid attack, and make sure it fails

### PR DESCRIPTION
**Pull request by cgolubi1**
I created this for use in reproducing the bug in #2989, and wanted to commit it.

It's not perfect, and it doesn't directly test any new cases since the fix for the outstanding issue should be pretty solid.  But it establishes a framework for expecting `submitTurn` to fail, and that's something i should be able to build on to do a lot more fuzz-testing of things that shouldn't work, so i'd like to get it in the codebase.

### CODE REVIEW

**Comment by blackshadowshade** [comment]
I acknowledge that this is random_ai only so I'm not going to block these changes. However, the code itself doesn't completely make sense to me, I'll add comments to let you know where.

File: tools/api-client/python/lib/random_ai.py [hunk 1]
```diff
@@ -1405,6 +1414,24 @@ def _should_allow_fire_cancellation(self):
...
+    funcPrefix = '_game_action_start_turn_find_attack_'
+    return [ x[len(funcPrefix):] for x in sorted(dir(self)) if x.startswith(funcPrefix) ]
+
+  # Given the list of valid attack types, choose one
+  # With low probability, choose an invalid attack type, and expect it to fail.
+  def _choose_attack_type(self, validAttackTypes):
+    invalidAttackTypes = self._all_known_attack_types()
+    if random.random() < 0.01:
+      invalidAttackType = self._random_array_element(invalidAttackTypes)
+      if invalidAttackType not in validAttackTypes:
```
Comment (blackshadowshade) [inline tools/api-client/python/lib/random_ai.py:1427]: So, I get that for some low probability (less than 1%), we'll get an invalid attack type here. However, I think that the code (or perhaps the variable names) are somewhat inaccurate.

invalidAttackTypes never actually ends up holding an array of invalid attack types. Nor is invalidAttackType guaranteed to hold an invalid attack type here.

If I were to rewrite this code to match the variable names, I'd expect to have to do a set subtraction of valid attack types from all attack types.

Comment (cgolubi1) [reply tools/api-client/python/lib/random_ai.py:1427]: Ah, you're right, i just need to change the variable name to `allAttackTypes` to make it match the logic (or do something like your suggestion, but that's probably more elegance than i need for this).  Good catch.

**Comment by cgolubi1** [comment]
> I acknowledge that this is random_ai only so I'm not going to block these changes. However, the code itself doesn't completely make sense to me, I'll add comments to let you know where.

New revision addresses this comment lazily, by making the variable names match the behavior and adding a comment to describe the behavior.

**Comment by blackshadowshade** [comment]
Yeah, that's better, thanks. Accepting now.

**Pull request merged**